### PR TITLE
feat(evm): Reuse geth code for `ApplyTransaction`

### PR DESCRIPTION
**Pull request by itsdevbear**

### CODE REVIEW

File: eth/core/processor.go [hunk 1]
```diff
@@ -140,94 +137,35 @@ func (sp *StateProcessor) Prepare(evm *vm.GethEVM, header *types.Header) {
...
-	receipt.Bloom = types.CreateBloom(types.Receipts{receipt})
-	receipt.TransactionIndex = uint(len(sp.txs))
-	receipt.BlockNumber = sp.header.Number
-
 	// Update the block information.
 	sp.txs = append(sp.txs, tx)
 	sp.receipts = append(sp.receipts, receipt)
 
 	// Return the execution result to the caller.
-	return result, err
```
Comment (calbera) [inline eth/core/processor.go:230]: lets add an empty blockhash to the receipt here?

File: eth/core/processor.go [hunk 2]
```diff
@@ -242,10 +180,12 @@ func (sp *StateProcessor) Finalize(
...
 	// Finalize the block with the txs and receipts (sets the TxHash, ReceiptHash, and Bloom).
 	block := types.NewBlock(sp.header, sp.txs, nil, sp.receipts, trie.NewStackTrie(nil))
-	blockHash := block.Hash()
 
-	// Set hashes on the receipts and logs.
-	var logs []*types.Log
+	// Update hashes on the receipts and logs, we have to do this in Finalize since when
+	// the transaction is actually being processed, we don't have the real blockhash. This is a
+	// fundamental design difference in the way polaris vs geth processes transactions.
+	logs := make([]*types.Log, 0)
```
Comment (calbera) [inline eth/core/processor.go:187]:
```suggestion
	var logs []*types.Log
```

File: eth/core/chain_writer.go [hunk 3]
```diff
@@ -156,15 +159,16 @@ func (bc *blockchain) Finalize(ctx context.Context) error {
...
-		bc.currentLogs.Store(logs)
-		if len(logs) > 0 {
-			bc.logsFeed.Send(logs)
+		if logs != nil {
+			bc.pendingLogsFeed.Send(logs)
+			bc.currentLogs.Store(logs)
+			if len(logs) > 0 {
+				bc.logsFeed.Send(logs)
+			}
 		}
```
Comment (calbera) [inline eth/core/chain_writer.go:172]: this should still be its own case and not nested in `receiptes != nil`

File: cosmos/testing/integration/graphql/graphql_test.go [hunk 4]
```diff
@@ -71,25 +70,25 @@ var _ = Describe("GraphQL", func() {
...
-					gasUsed
-				} 
-			} 
-		}`, addr.String(), calldata)
-		_, status, err := tf.SendGraphQLRequest(query)
-
-		Expect(status).To(Equal(200))
-		Expect(err).ToNot(HaveOccurred())
-	})
+	// It("should support eth_call", func() {
```
Comment (calbera) [inline cosmos/testing/integration/graphql/graphql_test.go:73]: whats wrong with this test?

File: cosmos/x/evm/keeper/msg_server.go [hunk 5]
```diff
@@ -41,21 +41,13 @@ func (k *Keeper) EthTransaction(
 	ctx context.Context, msg *types.EthTransactionRequest,
 ) (*types.EthTransactionResponse, error) {
 	// Process the transaction and return the result.
-	result, err := k.ProcessTransaction(ctx, msg.AsTransaction())
+	receipt, err := k.ProcessTransaction(ctx, msg.AsTransaction())
```
Comment (shampoobera) [inline cosmos/x/evm/keeper/msg_server.go:44]: ``ProcessTransaction`` returns an ``ExecutionResult``, maybe rename variable to ``executionResult`` for clarity

Comment (itsdevbear) [reply cosmos/x/evm/keeper/msg_server.go:44]: This PR changes it to return receipt

File: eth/core/processor_test.go [hunk 6]
```diff
@@ -80,7 +80,8 @@ var _ = Describe("StateProcessor", func() {
 			return false
 		}
 		gp.SetBlockGasLimit(uint64(blockGasLimit))
-
+		sdb.SetTxContextFunc = func(thash common.Hash, ti int) {}
+		sdb.TxIndexFunc = func() int { return 0 }
```
Comment (calbera) [inline eth/core/processor_test.go:84]: just add these on the mock?

File: eth/core/processor_test.go [hunk 7]
```diff
@@ -127,6 +127,7 @@ var _ = Describe("StateProcessor", func() {
 			sdb.GetBalanceFunc = func(addr common.Address) *big.Int {
 				return big.NewInt(1000001)
 			}
+			sdb.FinaliseFunc = func(bool) {}
```
Comment (calbera) [inline eth/core/processor_test.go:130]: +1

File: eth/core/processor_test.go [hunk 8]
```diff
@@ -159,6 +160,7 @@ var _ = Describe("StateProcessor", func() {
 			sdb.ExistFunc = func(addr common.Address) bool {
 				return addr == dummyContract
 			}
+			sdb.FinaliseFunc = func(bool) {}
```
Comment (calbera) [inline eth/core/processor_test.go:163]: +1

**Review by calbera — APPROVED** [review]
Lgtm, just remove the statedb mutex, will make tests pass too

**Comment by itsdevbear** [comment]
I don't think we need the mutex so gonna rip it out

**Pull request merged**